File: backend/src/services/schedule.service.ts
import { ScheduleRepository } from '../repositories/schedule.repository';
import { CatalogRepository } from '../repositories/catalog.repository';

export class ScheduleService {
  constructor(
    private scheduleRepository: ScheduleRepository,
    private catalogRepository: CatalogRepository
  ) {}

  async createAvulsoSession(data: {
    workspaceId: string;
    offeringId: string;
    professionalId: string;
    personId: string;
    sessionDate: Date;
    startMinute: number;
  }) {
    // Descobrir o tempo do Offering para calcular endMinute
    const offering = await this.catalogRepository.findOfferingById(data.offeringId);
    if (!offering) throw new Error('Offering não encontrado');

    const endMinute = data.startMinute + offering.durationMinutes;

    return this.scheduleRepository.createAvulsoSession({
      workspaceId: data.workspaceId,
      offeringId: data.offeringId,
      professionalId: data.professionalId,
      personId: data.personId,
      sessionDate: data.sessionDate,
      startMinute: data.startMinute,
      endMinute,
    });
  }

  async getSessions(workspaceId: string, startDate: Date, endDate: Date) {
    return this.scheduleRepository.getSessionsByDateRange(workspaceId, startDate, endDate);
  }
}

File: frontend/lib/services/schedule_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/schedule.dart';

class ScheduleService {
  final String baseUrl = 'http://localhost:3333/api';

  Future<List<Session>> getSessions(String workspaceId, DateTime startDate, DateTime endDate) async {
    try {
      final start = startDate.toIso8601String();
      final end = endDate.toIso8601String();
      final response = await http.get(
        Uri.parse('$baseUrl/schedule/sessions?workspaceId=$workspaceId&startDate=$start&endDate=$end'),
        headers: {'Content-Type': 'application/json', 'workspace-id': workspaceId},
      );
      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        return data.map((json) => Session.fromJson(json)).toList();
      }
    } catch (e) {
      print('Erro ao buscar sessões: $e');
    }
    return [];
  }

  Future<bool> createAvulsoSession(
    String workspaceId,
    String offeringId,
    String professionalId,
    String personId,
    DateTime sessionDate,
    int startMinute,
  ) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/schedule/sessions'),
        headers: {'Content-Type': 'application/json', 'workspace-id': workspaceId},
        body: jsonEncode({
          'workspaceId': workspaceId,
          'offeringId': offeringId,
          'professionalId': professionalId,
          'personId': personId,
          'sessionDate': sessionDate.toIso8601String(),
          'startMinute': startMinute,
        }),
      );
      return response.statusCode == 201;
    } catch (e) {
      print('Erro ao criar sessão avulsa: $e');
    }
    return false;
  }
}
